cache/remotecache/gha/gha.go

```diff
@@ -0,0 +1,347 @@
+	}
+	annotations := map[string]string{}
+	if l.Annotations.DiffID == "" {
+		return nil, errors.Errorf("cache layer with missing diffid")
+	}
+	annotations["containerd.io/uncompressed"] = l.Annotations.DiffID.String()
+	if !l.Annotations.CreatedAt.IsZero() {
+		txt, err := l.Annotations.CreatedAt.MarshalText()
+		if err != nil {
+			return nil, err
+		}
+		annotations["buildkit/createdat"] = string(txt)
+	}
+	desc := specs.Descriptor{
+		MediaType:   l.Annotations.MediaType,
+		Digest:      l.Blob,
+		Size:        l.Annotations.Size,
+		Annotations: annotations,
+	}
+	return &v1.DescriptorProviderPair{
+		Descriptor: desc,
+		Provider:   &ciProvider{desc: desc, ci: ci},
+	}, nil
+}
+
+func (ci *importer) loadScope(ctx context.Context, scope string) (*v1.CacheChains, error) {
+	scope = digest.FromBytes([]byte(scope)).Hex()[:8]
+	key := "index-" + ci.config.Scope + "-" + version + "-" + scope
+
+	entry, err := ci.cache.Load(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return v1.NewCacheChains(), nil
+	}
+
+	// TODO: this buffer can be removed
+	buf := &bytes.Buffer{}
+	if err := entry.WriteTo(ctx, buf); err != nil {
+		return nil, err
+	}
+
+	var config v1.CacheConfig
+	if err := json.Unmarshal(buf.Bytes(), &config); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	allLayers := v1.DescriptorProvider{}
+
+	for _, l := range config.Layers {
+		dpp, err := ci.makeDescriptorProviderPair(l)
+		if err != nil {
+			return nil, err
+		}
+		allLayers[l.Blob] = *dpp
+	}
+
+	cc := v1.NewCacheChains()
+	if err := v1.ParseConfig(config, allLayers, cc); err != nil {
+		return nil, err
+	}
+	return cc, nil
+}
+
+func (ci *importer) Resolve(ctx context.Context, _ specs.Descriptor, id string, w worker.Worker) (solver.CacheManager, error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	ccs := make([]*v1.CacheChains, len(ci.cache.Scopes()))
+
+	for i, s := range ci.cache.Scopes() {
+		func(i int, scope string) {
+			eg.Go(func() error {
+				cc, err := ci.loadScope(ctx, scope)
+				if err != nil {
+					return err
+				}
+				ccs[i] = cc
+				return nil
+			})
+		}(i, s.Scope)
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	cms := make([]solver.CacheManager, 0, len(ccs))
+
+	for _, cc := range ccs {
+		keysStorage, resultStorage, err := v1.NewCacheKeyStorage(cc, w)
+		if err != nil {
+			return nil, err
+		}
+		cms = append(cms, solver.NewCacheManager(id, keysStorage, resultStorage))
+	}
+
+	return solver.NewCombinedCacheManager(cms, nil), nil
+}
+
+type ciProvider struct {
+	desc specs.Descriptor
+	ci   *importer
+}
+
+func (p *ciProvider) ReaderAt(ctx context.Context, desc specs.Descriptor) (content.ReaderAt, error) {
+	key := "buildkit-blob-" + version + "-" + desc.Digest.String()
+	ce, err := p.ci.cache.Load(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if ce == nil {
+		return nil, errors.Errorf("blob not found")
+	}
+	rac := ce.Download(context.TODO())
```

Comment:
Can't we pass the ctx here?

Reply:
This context is long-running. Canceling it (allowed after this function returns) would cancel the `ReaderAt` that it returns. It should use the tracing context though, or the library should take a HTTP client so that tracing transport can be used.